app/forms/signals_telegram_bot_registration_form.rb

```diff
@@ -0,0 +1,37 @@
+class SignalsTelegramBotRegistrationForm < Patterns::Form
+  REQUIRED_MIN_STAKED_TOKEN_AMOUNT = 20000
+  param_key 'signals_telegram_bot'
+
+  attribute :telegram_username
+  attribute :chat_id
+  attribute :started_at
+
+  validates :telegram_username, presence: true
+  validates :user, presence: true, :if => proc { |f| f.telegram_username.present? }
+  validate :validate_user_staked_token_amount, :if => proc { |f| f.user.present? }
+  validates :chat_id, presence: true
+  validates :started_at, presence: true
+
+  def persist
+    user.token_sale[:signals_telegram_bot_chat_id] = chat_id
+    user.token_sale[:signals_telegram_bot_started_at] = started_at
+    user.save
+  end
+
+  protected
+
+  def user
+    @user ||= User.where("(token_sale->>'telegram_username') = ?", telegram_username).first
```

Comment:
Are we okay to just throw an error if no user is found with the provided `telegram_username`?

Reply:
Does it throw an error? It shouldn't and would just return `nil` if not found

Reply:
Sorry, this line won't throw an error but I'm thinking the `#persist` method would subsequently throw since we'd be calling `nil.token_sale` in the case of no user found.

Reply:
Ah, the logic behind this is in the parent `Patterns::Form` class. The `persist` method won't run until the validations pass and there is a validation for `validates : user, presence: true...`

Reply:
Ah very cool, thanks for the explanation!